crypto/src/memo.rs

```diff
@@ -18,64 +18,51 @@ pub const MEMO_CIPHERTEXT_LEN_BYTES: usize = 528;
 // This is the `MEMO_CIPHERTEXT_LEN_BYTES` - MAC size (16 bytes).
 pub const MEMO_LEN_BYTES: usize = 512;
 
-// The memo is stored separately from the `Note`.
-// TODO: MemoPlaintext should just be a fixed-length string, drop this type entirely
-#[derive(Clone, PartialEq, Eq)]
-pub struct MemoPlaintext(pub [u8; MEMO_LEN_BYTES]);
-
-impl Debug for MemoPlaintext {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        write!(f, "MemoPlaintext({})", hex::encode(self.0))
-    }
-}
-
-impl Default for MemoPlaintext {
-    fn default() -> MemoPlaintext {
-        MemoPlaintext([0u8; MEMO_LEN_BYTES])
-    }
-}
-
-impl TryFrom<&[u8]> for MemoPlaintext {
-    type Error = anyhow::Error;
-
-    fn try_from(input: &[u8]) -> Result<MemoPlaintext, Self::Error> {
-        if input.len() > MEMO_LEN_BYTES {
-            return Err(anyhow::anyhow!("provided memo exceeds maximum memo size"));
-        }
-        let mut mp = [0u8; MEMO_LEN_BYTES];
-        mp[..input.len()].copy_from_slice(input);
-
-        Ok(MemoPlaintext(mp))
-    }
-}
-
 #[derive(Clone, Debug)]
 pub struct MemoCiphertext(pub [u8; MEMO_CIPHERTEXT_LEN_BYTES]);
 
-impl MemoPlaintext {
+impl MemoCiphertext {
     /// Encrypt a memo, returning its ciphertext.
-    pub fn encrypt(&self, memo_key: PayloadKey) -> MemoCiphertext {
-        let encryption_result = memo_key.encrypt(self.0.to_vec(), PayloadKind::Memo);
+    pub fn encrypt(memo: &str, memo_key: PayloadKey) -> Result<MemoCiphertext, anyhow::Error> {
+        if memo.as_bytes().len() > MEMO_LEN_BYTES {
+            return Err(anyhow::anyhow!(
+                "provided memo plaintext exceeds maximum memo size"
+            ));
+        }
+        let mut m = [0u8; MEMO_LEN_BYTES];
+        m[..memo.as_bytes().len()].copy_from_slice(memo.as_bytes());
```

Comment:
Necessary! Otherwise:

```
---- memo::tests::test_memo_encryption_and_decryption stdout ----
thread 'memo::tests::test_memo_encryption_and_decryption' panicked at 'source slice length (2) does not match destination slice length (512)', crypto/src/memo.rs:34:15
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

However, I can still make it terser, by reusing a var for the `memo_len` value, since the error message needs that anyway.